File: typescript/tests/serializing.test.ts
import { describe, it, expect } from "vitest";
import { PrimaryModel, SecondaryModel } from "../output";

describe("Simple Fields — toJson, toRecord, toIRecord", async () => {
	const newRecord = new PrimaryModel({
		primaryKey: "Serialize Simple Test",
		singleLineText: "Hello World",
		longText: "Long text content",
		longTextWithRichText: "Rich text content",
		email: "test@example.com",
		url: "https://example.com",
		phoneNumber: "555-1234",
		checkbox: true,
		numberInt: 42,
		numberFloat: 3.14,
		currencyInt: 10,
		currencyFloat: 9.99,
		percentInt: 0.5,
		percentFloat: 0.333,
		duration: 3600,
		rating: 3,
		date: "2025-01-15",
		dateWithTime: "2025-01-15T10:00:00.000Z",
		singleSelect: "Choice 1",
		multipleSelect: ["Option 1", "Option 2"],
	});

	await newRecord.save();

	describe("toJson()", () => {
		const json = newRecord.toJson();

		it("should have id", () => {
			expect(json.id).toBe(newRecord.id);
		});

		it("should have all simple property values", () => {
			expect(json.primaryKey).toBe("Serialize Simple Test");
			expect(json.singleLineText).toBe("Hello World");
			expect(json.longText).toBe("Long text content");
			expect(json.longTextWithRichText).toBe("Rich text content");
			expect(json.email).toBe("test@example.com");
			expect(json.url).toBe("https://example.com");
			expect(json.phoneNumber).toBe("555-1234");
			expect(json.checkbox).toBe(true);
			expect(json.numberInt).toBe(42);
			expect(json.numberFloat).toBe(3.14);
			expect(json.currencyInt).toBe(10);
			expect(json.currencyFloat).toBe(9.99);
			expect(json.percentInt).toBe(0.5);
			expect(json.percentFloat).toBe(0.333);
			expect(json.duration).toBe(3600);
			expect(json.rating).toBe(3);
			expect(json.date).toBe("2025-01-15");
			expect(json.dateWithTime).toBe("2025-01-15T10:00:00.000Z");
			expect(json.singleSelect).toBe("Choice 1");
			expect(json.multipleSelect).toEqual(["Option 1", "Option 2"]);
		});

		it("should have linked record fields as raw IDs (not LinkedRecord objects)", () => {
			// linkSingle and linkMultiple were not set, so they should be undefined/empty
			expect(json.linkSingle).toBeUndefined();
			expect(json.linkMultiple).toBeUndefined();
		});
	});

	describe("toRecord()", () => {
		it("should return a record with matching id", () => {
			const r = newRecord.toRecord();
			expect(r.id).toBe(newRecord.id);
		});

		it("should use field names as keys by default", () => {
			const r = newRecord.toRecord();
			expect(r.fields["Primary Key"]).toBe("Serialize Simple Test");
			expect(r.fields["Single Line Text"]).toBe("Hello World");
			expect(r.fields["Checkbox"]).toBe(true);
			expect(r.fields["Number (int)"]).toBe(42);
			expect(r.fields["Date"]).toBe("2025-01-15");
			expect(r.fields["Single Select"]).toBe("Choice 1");
			expect(r.fields["Multiple Select"]).toEqual(["Option 1", "Option 2"]);
		});

		it("should use field IDs as keys when useFieldIds=true", () => {
			const r = newRecord.toRecord(true);
			// "Primary Key" field ID is "fldol5Q4wmQJQvPRy"
			expect((r.fields as any)["fldol5Q4wmQJQvPRy"]).toBe("Serialize Simple Test");
			// "Single Line Text" field ID is "fld0BL2lFo9fqcKv3"
			expect((r.fields as any)["fld0BL2lFo9fqcKv3"]).toBe("Hello World");
			// "Checkbox" field ID is "fldjQIaAZVegb1FUa"
			expect((r.fields as any)["fldjQIaAZVegb1FUa"]).toBe(true);
		});
	});

	describe("toIRecord()", () => {
		it("should return an object with only id and fields", () => {
			const ir = newRecord.toIRecord();
			expect(ir.id).toBe(newRecord.id);
			expect(ir.fields).toBeDefined();
			// Should only have id and fields (no ATRecord methods)
			const keys = Object.keys(ir);
			expect(keys).toContain("id");
			expect(keys).toContain("fields");
			expect(keys.length).toBe(2);
		});

		it("should use field names as keys by default", () => {
			const ir = newRecord.toIRecord();
			expect(ir.fields["Primary Key"]).toBe("Serialize Simple Test");
			expect(ir.fields["Single Line Text"]).toBe("Hello World");
			expect(ir.fields["Number (int)"]).toBe(42);
		});

		it("should use field IDs as keys when useFieldIds=true", () => {
			const ir = newRecord.toIRecord(true);
			expect((ir.fields as any)["fldol5Q4wmQJQvPRy"]).toBe("Serialize Simple Test");
			expect((ir.fields as any)["fld0BL2lFo9fqcKv3"]).toBe("Hello World");
		});
	});

	describe("Cleanup", () => {
		it("should delete the record", async () => {
			await newRecord.delete();
		});
	});
});

describe("Linked Records — toJson, toRecord, toIRecord", async () => {
	const sec1 = new SecondaryModel({ name: "Ser Link 1", value: "val1" });
	await sec1.save();
	const sec2 = new SecondaryModel({ name: "Ser Link 2", value: "val2" });
	await sec2.save();
	const sec1Id = sec1.id!;
	const sec2Id = sec2.id!;

	const newRecord = new PrimaryModel({
		primaryKey: "Serialize Link Test",
		linkSingle: sec1Id,
		linkMultiple: [sec1Id, sec2Id],
	});
	await newRecord.save();

	describe("toJson()", () => {
		const json = newRecord.toJson();

		it("should have linkSingle as a single record ID string", () => {
			expect(json.linkSingle).toBe(sec1Id);
		});

		it("should have linkMultiple as an array of record ID strings", () => {
			expect(json.linkMultiple).toEqual([sec1Id, sec2Id]);
		});
	});

	describe("toRecord()", () => {
		it("should have linked fields as arrays of record IDs with field name keys", () => {
			const r = newRecord.toRecord();
			// In Airtable's native format, single link is [id], multiple is [id1, id2]
			expect(r.fields["Link (single)"]).toEqual([sec1Id]);
			expect(r.fields["Link (multiple)"]).toEqual([sec1Id, sec2Id]);
		});
	});

	describe("toIRecord()", () => {
		it("should have linked fields as arrays of record IDs with { id, fields } shape", () => {
			const ir = newRecord.toIRecord();
			expect(ir.fields["Link (single)"]).toEqual([sec1Id]);
			expect(ir.fields["Link (multiple)"]).toEqual([sec1Id, sec2Id]);
			// Confirm only id and fields
			const keys = Object.keys(ir);
			expect(keys).toContain("id");
			expect(keys).toContain("fields");
			expect(keys.length).toBe(2);
		});
	});

	describe("Cleanup", () => {
		it("should clean up records", async () => {
			await newRecord.delete();
			await sec1.delete();
			await sec2.delete();
		});
	});
});

describe("Unfetched model", async () => {
	describe("toJson() on a new model before saving", () => {
		it("should work and return the local data", () => {
			const m = new PrimaryModel({ primaryKey: "Unsaved Model" });
			const json = m.toJson();
			expect(json.primaryKey).toBe("Unsaved Model");
			expect(json.id).toBeUndefined();
		});
	});

	describe("toRecord() on a new model before saving", () => {
		it("should work since the constructor creates the internal record", () => {
			const m = new PrimaryModel({ primaryKey: "Unsaved Model" });
			const r = m.toRecord();
			expect(r.fields["Primary Key"]).toBe("Unsaved Model");
		});
	});

	describe("toRecord() on fromId without fetch", async () => {
		// Create a real record to get a valid ID
		const temp = new PrimaryModel({ primaryKey: "FromId No Fetch" });
		await temp.save();
		const tempId = temp.id!;

		it("should work but fields will be empty (no field data fetched)", () => {
			const m = PrimaryModel.fromId(tempId);
			const r = m.toRecord();
			expect(r.id).toBe(tempId);
			// Fields should be empty since we never called fetch()
			expect(r.fields["Primary Key"]).toBeUndefined();
		});

		it("cleanup", async () => {
			await temp.delete();
		});
	});
});

File: go/tests/serializing_test.go
package tests

// G4.6 — Offline (no-network) model serialization parity with Java TestSerializing /
// Kotlin TestSerializing / Swift TestSerializing. These cases exercise the EXPORTED
// generated model structs with stdlib encoding/json directly: the json tags are the
// Airtable field IDs, so a fields-object keyed by field ID round-trips through the
// model. No credentials or network are required — these must run, not skip.

import (
	"context"
	"encoding/json"
	"testing"

	airtable "myairtabletests/output"
)

// decodePrimary unmarshals a field-ID-keyed JSON object into a PrimaryModel.
func decodePrimary(t *testing.T, body string) airtable.PrimaryModel {
	t.Helper()
	var m airtable.PrimaryModel
	if err := json.Unmarshal([]byte(body), &m); err != nil {
		t.Fatalf("unmarshal PrimaryModel: %v", err)
	}
	return m
}

func TestSerializing(t *testing.T) {
	t.Run("DecodeWritableFieldsKeyedByFieldID", func(t *testing.T) {
		body := `{
			"` + airtable.PrimaryPrimaryKeyFieldID + `": "PK Value",
			"` + airtable.PrimaryEmailFieldID + `": "a@b.c",
			"` + airtable.PrimaryNumberIntFieldID + `": 42,
			"` + airtable.PrimaryCheckboxFieldID + `": true
		}`
		m := decodePrimary(t, body)
		if m.PrimaryKey == nil || *m.PrimaryKey != "PK Value" {
			t.Fatalf("PrimaryKey = %v, want \"PK Value\"", m.PrimaryKey)
		}
		if m.Email == nil || *m.Email != "a@b.c" {
			t.Fatalf("Email = %v, want \"a@b.c\"", m.Email)
		}
		if m.NumberInt == nil || *m.NumberInt != 42 {
			t.Fatalf("NumberInt = %v, want 42", m.NumberInt)
		}
		if m.Checkbox == nil || *m.Checkbox != true {
			t.Fatalf("Checkbox = %v, want true", m.Checkbox)
		}
	})

	t.Run("AbsentFieldIsNilNotZeroValue", func(t *testing.T) {
		m := decodePrimary(t, `{"`+airtable.PrimaryPrimaryKeyFieldID+`": "only pk"}`)
		if m.PrimaryKey == nil || *m.PrimaryKey != "only pk" {
			t.Fatalf("PrimaryKey = %v, want \"only pk\"", m.PrimaryKey)
		}
		// Absent writable fields decode to nil pointers, distinguishing "unset"
		// from a zero value ("" / 0 / false).
		if m.Email != nil {
			t.Fatalf("Email = %v, want nil for absent field", m.Email)
		}
		if m.NumberInt != nil {
			t.Fatalf("NumberInt = %v, want nil for absent field", m.NumberInt)
		}
		if m.Checkbox != nil {
			t.Fatalf("Checkbox = %v, want nil for absent field", m.Checkbox)
		}
	})

	t.Run("ComputedFieldDecodesPlainValue", func(t *testing.T) {
		// Auto Number is a computed *MaybeSpecialOrError[int64]; a plain number
		// decodes to a present value.
		m := decodePrimary(t, `{"`+airtable.PrimaryAutoNumberFieldID+`": 7}`)
		if m.AutoNumber == nil {
			t.Fatal("AutoNumber = nil, want present wrapper")
		}
		v, ok := m.AutoNumber.Value()
		if !ok || v != 7 {
			t.Fatalf("AutoNumber.Value() = (%d, %v), want (7, true)", v, ok)
		}
		if m.AutoNumber.IsSpecial() || m.AutoNumber.IsError() {
			t.Fatal("plain value should be neither special nor error")
		}
	})

	t.Run("ComputedFieldDecodesSpecialValue", func(t *testing.T) {
		m := decodePrimary(t, `{"`+airtable.PrimaryAutoNumberFieldID+`": {"specialValue": "NaN"}}`)
		if m.AutoNumber == nil {
			t.Fatal("AutoNumber = nil, want present wrapper")
		}
		if !m.AutoNumber.IsSpecial() {
			t.Fatal("AutoNumber.IsSpecial() = false, want true")
		}
		if _, ok := m.AutoNumber.Value(); ok {
			t.Fatal("AutoNumber.Value() ok = true for a special number, want false")
		}
		if sn := m.AutoNumber.Special(); sn == nil || sn.SpecialValue != "NaN" {
			t.Fatalf("Special() = %v, want SpecialValue=NaN", sn)
		}
	})

	t.Run("ComputedFieldDecodesErrorValue", func(t *testing.T) {
		m := decodePrimary(t, `{"`+airtable.PrimaryAutoNumberFieldID+`": {"error": "#ERROR!"}}`)
		if m.AutoNumber == nil {
			t.Fatal("AutoNumber = nil, want present wrapper")
		}
		if !m.AutoNumber.IsError() {
			t.Fatal("AutoNumber.IsError() = false, want true")
		}
		if _, ok := m.AutoNumber.Value(); ok {
			t.Fatal("AutoNumber.Value() ok = true for an error, want false")
		}
		if ev := m.AutoNumber.ErrorVal(); ev == nil || ev.Error != "#ERROR!" {
			t.Fatalf("ErrorVal() = %v, want Error=#ERROR!", ev)
		}
	})

	t.Run("AirtableTimeDecodesDateOnlyAndRFC3339", func(t *testing.T) {
		m := decodePrimary(t, `{
			"`+airtable.PrimaryDateFieldID+`": "2024-01-15",
			"`+airtable.PrimaryDateWithTimeFieldID+`": "2024-01-15T10:30:00.000Z"
		}`)
		if m.Date == nil {
			t.Fatal("Date = nil, want decoded date-only value")
		}
		if got := m.Date.Time.UTC().Format("2006-01-02T15:04:05Z07:00"); got != "2024-01-15T00:00:00Z" {
			t.Fatalf("Date = %q, want 2024-01-15T00:00:00Z", got)
		}
		if m.DateWithTime == nil {
			t.Fatal("DateWithTime = nil, want decoded RFC3339 value")
		}
		if got := m.DateWithTime.Time.UTC().Format("2006-01-02T15:04:05Z07:00"); got != "2024-01-15T10:30:00Z" {
			t.Fatalf("DateWithTime = %q, want 2024-01-15T10:30:00Z", got)
		}
	})

	t.Run("AirtableDurationDecodesNumericSeconds", func(t *testing.T) {
		m := decodePrimary(t, `{"`+airtable.PrimaryDurationFieldID+`": 3661}`)
		if m.Duration == nil {
			t.Fatal("Duration = nil, want decoded numeric seconds")
		}
		if got := m.Duration.Duration().Seconds(); got != 3661 {
			t.Fatalf("Duration = %v seconds, want 3661", got)
		}
	})

	t.Run("SliceFieldDecodes", func(t *testing.T) {
		// Link (multiple) is a []string of linked record IDs.
		m := decodePrimary(t, `{"`+airtable.PrimaryLinkMultipleFieldID+`": ["recAAA", "recBBB"]}`)
		if len(m.LinkMultiple) != 2 || m.LinkMultiple[0] != "recAAA" || m.LinkMultiple[1] != "recBBB" {
			t.Fatalf("LinkMultiple = %v, want [recAAA recBBB]", m.LinkMultiple)
		}
	})

	t.Run("MarshalEmitsFieldIDKeysAndOmitsNil", func(t *testing.T) {
		m := airtable.PrimaryModel{
			PrimaryKey: airtable.String("x"),
			Email:      airtable.String("a@b.c"),
			NumberInt:  airtable.Float64(7),
		}
		encoded, err := json.Marshal(&m)
		if err != nil {
			t.Fatalf("marshal: %v", err)
		}
		// Re-decode into a generic map so assertions are key-order independent.
		var out map[string]json.RawMessage
		if err := json.Unmarshal(encoded, &out); err != nil {
			t.Fatalf("re-unmarshal: %v", err)
		}
		// Set writable pointer fields appear under their field-ID keys.
		if string(out[airtable.PrimaryPrimaryKeyFieldID]) != `"x"` {
			t.Fatalf("PrimaryKey key = %s, want \"x\"", out[airtable.PrimaryPrimaryKeyFieldID])
		}
		if string(out[airtable.PrimaryEmailFieldID]) != `"a@b.c"` {
			t.Fatalf("Email key = %s, want \"a@b.c\"", out[airtable.PrimaryEmailFieldID])
		}
		if string(out[airtable.PrimaryNumberIntFieldID]) != `7` {
			t.Fatalf("NumberInt key = %s, want 7", out[airtable.PrimaryNumberIntFieldID])
		}
		// Nil writable fields are omitted (omitempty on the pointer tag).
		if _, ok := out[airtable.PrimaryCheckboxFieldID]; ok {
			t.Fatal("Checkbox present in output, want omitted for nil pointer")
		}
		if _, ok := out[airtable.PrimaryDurationFieldID]; ok {
			t.Fatal("Duration present in output, want omitted for nil pointer")
		}
	})

	t.Run("UnsavedModelHasEmptyID", func(t *testing.T) {
		// A freshly constructed model is "unsaved": ID() is "" until create/hydrate.
		m := &airtable.PrimaryModel{PrimaryKey: airtable.String("never saved")}
		if m.ID() != "" {
			t.Fatalf("ID() = %q, want \"\" for an unsaved model", m.ID())
		}
		if m.CreatedTime() != nil {
			t.Fatalf("CreatedTime() = %v, want nil for an unsaved model", m.CreatedTime())
		}
	})
}

// TestSerializingIntegration — G10.1 network-bound serialization parity with the
// Java TestSerializingIntegration / Kotlin TestSerializingIntegration: linked-
// record ID arrays round-trip, createdTime is populated on create, unsaved
// models report an empty ID, and a writable field cleared to nil is unset
// server-side. Live; distinct primary keys; cleans up everything it creates.
func TestSerializingIntegration(t *testing.T) {
	at := newAirtable(t)
	ctx := context.Background()

	t.Run("CreatedTimeAndIDRoundTrip", func(t *testing.T) {
		pk := primaryKey("Serializing", "IdTime")
		created, err := at.Primary.CreateOne(ctx, &airtable.PrimaryModel{PrimaryKey: airtable.String(pk)})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer at.Primary.DeleteOne(ctx, id) //nolint:errcheck

		if id == "" {
			t.Fatal("created model has empty ID")
		}
		if created.CreatedTime() == nil {
			t.Fatal("CreatedTime() is nil on a created record, want populated")
		}

		fetched, err := at.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("get: %v", err)
		}
		if fetched.ID() != id {
			t.Fatalf("fetched ID = %s, want %s", fetched.ID(), id)
		}
		if fetched.CreatedTime() == nil {
			t.Fatal("fetched CreatedTime() is nil, want populated")
		}
		if !created.CreatedTime().Time.Equal(fetched.CreatedTime().Time) {
			t.Fatalf("createdTime mismatch: created %v vs fetched %v",
				created.CreatedTime().Time, fetched.CreatedTime().Time)
		}
	})

	t.Run("LinkedRecordIDArraysRoundTrip", func(t *testing.T) {
		suite := primaryKey("Serializing", "Links")
		sec1, err := at.Secondary.CreateOne(ctx, &airtable.SecondaryModel{Name: airtable.String(suite + " T1")})
		if err != nil {
			t.Fatalf("create secondary 1: %v", err)
		}
		defer at.Secondary.DeleteOne(ctx, sec1.ID()) //nolint:errcheck
		sec2, err := at.Secondary.CreateOne(ctx, &airtable.SecondaryModel{Name: airtable.String(suite + " T2")})
		if err != nil {
			t.Fatalf("create secondary 2: %v", err)
		}
		defer at.Secondary.DeleteOne(ctx, sec2.ID()) //nolint:errcheck

		prim, err := at.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:   airtable.String(suite),
			LinkSingle:   []string{sec1.ID()},
			LinkMultiple: []string{sec1.ID(), sec2.ID()},
		})
		if err != nil {
			t.Fatalf("create primary: %v", err)
		}
		defer at.Primary.DeleteOne(ctx, prim.ID()) //nolint:errcheck

		// Links serialize as []string of record IDs and round-trip through create.
		if len(prim.LinkSingle) != 1 || prim.LinkSingle[0] != sec1.ID() {
			t.Fatalf("LinkSingle = %v, want [%s]", prim.LinkSingle, sec1.ID())
		}
		if len(prim.LinkMultiple) != 2 {
			t.Fatalf("LinkMultiple len = %d, want 2", len(prim.LinkMultiple))
		}

		// Re-fetch and confirm the arrays survive a server round-trip.
		fetched, err := at.Primary.GetOne(ctx, prim.ID())
		if err != nil {
			t.Fatalf("get primary: %v", err)
		}
		if len(fetched.LinkSingle) != 1 || fetched.LinkSingle[0] != sec1.ID() {
			t.Fatalf("fetched LinkSingle = %v, want [%s]", fetched.LinkSingle, sec1.ID())
		}
		if len(fetched.LinkMultiple) != 2 {
			t.Fatalf("fetched LinkMultiple len = %d, want 2", len(fetched.LinkMultiple))
		}
	})

	t.Run("NullFieldClearingUnsetsServerSide", func(t *testing.T) {
		pk := primaryKey("Serializing", "Clear")
		created, err := at.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey:     airtable.String(pk),
			SingleLineText: airtable.String("to be cleared"),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer at.Primary.DeleteOne(ctx, id) //nolint:errcheck

		if created.SingleLineText == nil || *created.SingleLineText != "to be cleared" {
			t.Fatalf("created text = %v, want 'to be cleared'", created.SingleLineText)
		}

		// Clear the field (set the pointer to nil) and update: the dirty diff sends
		// JSON null, unsetting the field server-side.
		created.SingleLineText = nil
		saved, err := at.Primary.UpdateOne(ctx, created)
		if err != nil {
			t.Fatalf("update (clear): %v", err)
		}
		if saved.SingleLineText != nil {
			t.Fatalf("text after clear = %v, want nil", saved.SingleLineText)
		}

		fetched, err := at.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("get after clear: %v", err)
		}
		if fetched.SingleLineText != nil {
			t.Fatalf("fetched text after clear = %v, want nil (cleared server-side)", fetched.SingleLineText)
		}
	})
}
